### DISCUSSION TITLE
How does one get list of modulators in a preset?

### DISCUSSION THREAD

**Discussion by albedozero** [FluidSynth/fluidsynth#719] [Q&A] [answered]
I'm working on some code using Python's ctypes to bind to various Fluidsynth functions. I'm looking for a way to get the list of modulators in a preset, or perhaps the list of instrument+preset modulators.

It seems like they might be in instrument zone and preset zone data structures that are pointed to somewhere in the _data_ member of a _fluid_preset_t_, and could perhaps be accessed using the _fluid_preset_get_data_ function in _fluid_sfont.c_, but I'm having trouble unraveling what _fluid_preset_t->data_ would look like.

Any help on how to do this or whether it is, in general, possible?

**Comment by mawe42** [comment, marked as answer]
I don't think we provide a public API to retrieve the modulators of a preset. That knowledge is contained inside the private implementations of the soundfont loader. Depending on your needs, it might be easier to use a library like [libinstpatch](https://github.com/swami/libinstpatch) to retrieve that information from a Soundfont.

**Reply by albedozero** [reply]
That's what I gathered, looking through the code. Thanks for the suggestion!
